### PR TITLE
feat: Reintroduce collaboration feature

### PR DESCRIPTION
**Pull request by tomi**
## Summary

Collaboration feature [was removed](https://github.com/n8n-io/n8n/pull/9929) because it was missing authorization checks. This adds checks that user can't send workflow opened or closed message, unless they have access to that workflow. Also uses `cache` for storing the collaboration state, so the feature works correctly with multiple mains.

The feature is currently only enabled for the new canvas, as that should come live soon anyways.

## Related Linear tickets, Github issues, and Community forum posts

[CAT-79](https://linear.app/n8n/issue/CAT-79/bug-collaborationservice-does-not-implement-access-control)

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/cli/src/collaboration/collaboration.service.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		private readonly userService: UserService,
+		private readonly sharedWorkflowRepository: SharedWorkflowRepository,
+	) {}
+
+	init() {
+		this.push.on('message', async (event: OnPushMessage) => {
+			try {
+				await this.handleUserMessage(event.userId, event.msg);
+			} catch (error) {
+				this.logger.error('Error handling user message', {
```
Comment (tomi) [inline packages/cli/src/collaboration/collaboration.service.ts:35]: Should we report these to sentry? What's the correct way to do that? Throw an `ApplicationError` and let the top-level handler take care of it?

File: packages/cli/src/collaboration/collaboration.service.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		const user = await this.userRepository.findOneBy({
+			id: userId,
+		});
+		if (!user) {
+			return false;
+		}
+
+		const workflow = await this.sharedWorkflowRepository.findWorkflowForUser(workflowId, user, [
+			'workflow:read',
+		]);
```
Comment (tomi) [inline packages/cli/src/collaboration/collaboration.service.ts:111]: Ideally would want to check the permissions without needing to query the entire workflow. Might do that as a follow up

File: packages/cli/src/push/websocket.push.ts [hunk 3]
```diff
@@ -16,17 +17,34 @@ export class WebSocketPush extends AbstractPush<WebSocket> {
...
-		super.add(pushRef, connection);
+		super.add(pushRef, userId, connection);
+
+		const onMessage = (data: WebSocket.RawData) => {
+			try {
+				const buffer = Array.isArray(data) ? Buffer.concat(data) : Buffer.from(data);
+
+				this.onMessageReceived(pushRef, JSON.parse(buffer.toString('utf8')));
+			} catch (error) {
+				this.logger.error("Couldn't parse message from editor-UI", {
```
Comment (tomi) [inline packages/cli/src/push/websocket.push.ts:32]: Maybe should also report this to sentry?

File: packages/editor-ui/src/components/MainHeader/CollaborationPane.vue [hunk 4]
```diff
@@ -0,0 +1,82 @@
+<script setup lang="ts">
```
Comment (tomi) [inline packages/editor-ui/src/components/MainHeader/CollaborationPane.vue:1]: 1:1 copy from the one we had before

File: packages/editor-ui/src/views/NodeView.v2.vue [hunk 5]
```diff
@@ -252,6 +254,8 @@ async function initializeData() {
 
 	try {
 		await Promise.all(loadPromises);
+
+		collaborationStore.notifyWorkflowOpened(workflowId.value);
```
Comment (tomi) [inline packages/editor-ui/src/views/NodeView.v2.vue:258]: Is this the correct place where this should be sent?

File: packages/cli/src/collaboration/__tests__/collaboration.state.test.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+			// Act
+			const users = await collaborationState.getActiveWorkflowUsers(workflowId);
+
+			// Assert
+			expect(users).toEqual([
+				{
+					lastSeen: now,
+					userId: 'notExpiredUserId',
+				},
+			]);
```
Comment (despairblue) [inline packages/cli/src/collaboration/__tests__/collaboration.state.test.ts:90]: **suggestion:** Assert that expired users are removed from the cache:

```suggestion
			expect(users).toEqual([
				{
					lastSeen: now,
					userId: 'notExpiredUserId',
				},
			]);
			// removes expired users from the cache
			expect(mockCacheService.deleteFromHash).toHaveBeenCalledWith(
				'collaboration:workflow',
				'expiredUserId',
			);
```

File: packages/cli/src/push/__tests__/websocket.push.test.ts [hunk 7]
```diff
@@ -38,14 +40,14 @@ describe('WebSocketPush', () => {
 	});
 
 	it('can add a connection', () => {
-		webSocketPush.add(pushRef1, mockWebSocket1);
+		webSocketPush.add(pushRef1, userId, mockWebSocket1);
 
 		expect(mockWebSocket1.listenerCount('close')).toBe(1);
 		expect(mockWebSocket1.listenerCount('pong')).toBe(1);
```
Comment (despairblue) [inline packages/cli/src/push/__tests__/websocket.push.test.ts:46]:
```suggestion
		expect(mockWebSocket1.listenerCount('pong')).toBe(1);
		expect(mockWebSocket1.listenerCount('message')).toBe(1);
```

File: packages/cli/src/push/__tests__/websocket.push.test.ts [hunk 8]
```diff
@@ -38,14 +40,14 @@ describe('WebSocketPush', () => {
...
 	});
 
 	it('closes a connection', () => {
-		webSocketPush.add(pushRef1, mockWebSocket1);
+		webSocketPush.add(pushRef1, userId, mockWebSocket1);
 
 		mockWebSocket1.emit('close');
```
Comment (despairblue) [inline packages/cli/src/push/__tests__/websocket.push.test.ts:53]: **suggestion:** Assert `expect(mockWebSocket1.listenerCount('message')).toBe(0)`

This requires to remove all listeners in the before each:

```ts
	beforeEach(() => {
		jest.resetAllMocks();
		mockWebSocket1.removeAllListeners();
		mockWebSocket2.removeAllListeners();
	});
```

File: packages/cli/test/integration/collaboration/collaboration.service.test.ts [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+		});
+
+		it("should not emit activeWorkflowUsersChanged if user don't have access to the workflow", async () => {
+			// Arrange
+			const sendToUsersSpy = jest.spyOn(pushService, 'sendToUsers');
+			await sendWorkflowOpenedMessage(workflow.id, owner.id);
+			sendToUsersSpy.mockClear();
+
+			// Act
+			await sendWorkflowClosedMessage(workflow.id, owner.id);
```
Comment (despairblue) [inline packages/cli/test/integration/collaboration/collaboration.service.test.ts:182]:
```suggestion
			await sendWorkflowClosedMessage(workflow.id, memberWithoutAccess.id);

```

**Review by despairblue — CHANGES_REQUESTED** [review]
Disclaimer: I did not review the FE code.

The backend good looks good, also the permission related code. 🚀 

I only made some smaller suggestions to the tests.

File: packages/cli/src/collaboration/collaboration.message.ts
[the remark this answers is not in the document]
Comment (tomi) [reply packages/cli/src/collaboration/collaboration.message.ts:1]: Great suggestion! It's always better to parse any user input

**Comment by tomi** [comment]
Thank you for a great review @despairblue ! I addressed all your comments. LMK if I missed something ✌️

**Review by despairblue — DISMISSED** [review]
Awesome 🚀 

Approving the backend changes!

File: packages/editor-ui/src/stores/collaboration.store.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	});
+
+	function initialize() {
+		if (pushStoreEventListenerRemovalFn.value) {
+			return;
+		}
+
+		pushStoreEventListenerRemovalFn.value = pushStore.addEventListener((event) => {
+			if (event.type === 'activeWorkflowUsersChanged') {
+				console.log('Received event', event);
```
Comment (MiloradFilipovic) [inline packages/editor-ui/src/stores/collaboration.store.ts:36]: Should we remove these `console.log`s

File: packages/editor-ui/src/views/NodeView.v2.vue (hunk 5 above)
Comment (MiloradFilipovic) [reply packages/editor-ui/src/views/NodeView.v2.vue:267]: Maybe a better place would be the `initializeWorkspaceForExistingWorkflow` or `openWorkflow` method in the same file.

**Review by MiloradFilipovic — COMMENTED** [review]
Apart from the minor comments I left, front-end part looks good to me.
However, I am seeing some unexpected behavior when switching users (the list is not always synced in both user sessions):

https://github.com/user-attachments/assets/fa1b0e5c-c481-4d2c-96d4-f5f4d8b837da

**Comment by tomi** [comment]
> Apart from the minor comments I left, front-end part looks good to me. However, I am seeing some unexpected behavior when switching users (the list is not always synced in both user sessions):
> Screen.Recording.2024-08-30.at.18.00.39.mov

Couldn't reproduce anymore after changing the place where we notify workflow opened

**Review by MiloradFilipovic — APPROVED** [review]
Thanks for addressing!
Looking good with the latest changes 🙌

**Comment by janober** [comment]
Got released with [`n8n@1.58.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.58.0)

**Pull request merged**